### PR TITLE
Major improvements in safety and usability

### PR DESCRIPTION
**Pull request by jeremiahsnapp**
The goal of this PR is to make the knife-acl plugin a more powerful tool to manage Chef Server's Access Control Lists (ACLs) in a way that is safe.

The following changes address the safety concerns surrounding the incomplete implementation of USAGs:

1. Add a user to a group or ACE by the user's name instead of its USAG
2. Do not permit the management of the `users` group membership since this conflicts with the `users` group's default use of USAGs
3. Do not allow the attempted destruction of `admins`, `billing-admins`, `clients` or `users` group

The following makes knife-acl more powerful:

1. Remove the need for the actor-map.yaml file
2. New `acl bulk add` and `acl bulk remove` commands use REGEX to modify ACLs of matching objects (e.g. nodes, clients, etc)
3. Multiple ACEs (i.e. create, read, update, delete, grant) of an ACL can be modified using a single command
4. Allow a user to be added/removed from an ACL
5. Only make API requests when necessary to improve performance

### CODE REVIEW

**Comment by stevendanna** [comment]
I haven't had a chance to look through the code in detail yet but:

Changes I like:

- Adding users directly to the groups. 
- Moving the group add actor commands into two separate commands(add user, add client)
- Allowing multiple permissions to be updated with a single command
- The bulk commands in general, this will make setting up permissions schemes in existing orgs easier.

Changes I'm less sure on:

- Completely removing the actor-map.  Won't knife group show for the groups that contain USAGs become a lot less useful. 
- Disallowing interaction with the user group and other groups.  Perhaps we can fail by default but have a --force flag?

Given the amount of work you've put into this, I'm wondering if now is the time to bite the bullet and get some of this into core chef.

**Comment by jeremiahsnapp** [comment]
Thanks for taking a look @stevendanna and for the feedback.

> Completely removing the actor-map. Won't knife group show for the groups that contain USAGs become a lot less useful.

The only group that should have USAGs as far as I know is the Users group which should include all members of the organization by default. My opinion is that we should discourage any use of the Users group at least until USAGs get removed completely. So the `group show` should only be less useful for the Users group which we shouldn't really care about anyway.

> Disallowing interaction with the Users group and other groups. Perhaps we can fail by default but have a --force flag?

I believe the only thing I'm not allowing for "other groups" is the destruction of them. That goes for `admins`, `billing-admins`, `clients`, `users`. I consider these system groups and I don't know of any good reason why anyone should be destroying them (if it's even possible). In an effort to make knife-acl safe I feel it is reasonable to not allow knife-acl to perform this destruction at all. If someone really wants to try then they can read the code and make their own API requests but I doubt many will want to do this. Also, I'd rather handle a support ticket for their inability to use knife-acl to destroy the group instead of a ticket saying they used `--force` and now something's broken. :)

The other thing I'm preventing is the ability to add/remove members from the Users group. My opinion is that the Users group shouldn't be manipulated at all in order to avoid unnecessary trouble with things like the Manage UI. Read the older version of the knife-acl's README.md for more information about the conflicts that occur when trying to manage Users group with both knife-acl and Manage UI. Avoiding this altogether makes knife-acl much more safe to use for, hopefully, the majority of use cases.

If the user wants to restrict access to things then they should remove the Users group from all appropriate ACLS and then create a new group with the appropriate actors and add that new group to the appropriate ACLs.

File: lib/chef/knife/acl_bulk_add.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      end
+
+      find_actor(actor_type, actor_name)
+
+      objects_to_modify = []
+      all_objects = rest.get_rest(object_type)
+      all_objects.keys.each do |object_name|
+        next unless object_name =~ object_name_matcher
+        objects_to_modify << object_name
+      end
```
Comment (stevendanna) [inline lib/chef/knife/acl_bulk_add.rb]: I'm not sure if it "more idiomatic", but an alternative I like is:

```
all_objects = rest.get_rest(object_type)
objects_to_modify = all_objects.keys.select { |object_name| object_name =~ object_name_matcher }
```

File: lib/chef/knife/acl_bulk_add.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
+#
+# Author:: Steven Danna (steve@opscode.com)
```
Comment (stevendanna) [inline lib/chef/knife/acl_bulk_add.rb]: Might want to fix up the copyright & author notices.

File: lib/chef/knife/acl_remove.rb [hunk 3]
```diff
@@ -19,45 +19,27 @@
...
-      when "users"
-        # Not Implemented yet, we shouldn't get here.
-      end
-
-      update_ace!(object_type, object_name, perm, ace)
-    end
-
-    def remove_group_from_ace!(name, ace)
-      ace['groups'].delete(name)
+      find_actor(actor_type, actor_name)
```
Comment (stevendanna) [inline lib/chef/knife/acl_remove.rb]: It seems that we are always ignoring the return value of this function.  Are we only ever using it to fail early if the given actor doesn't exist?  If so, you might want to give this a name similar to the other "validate" functions.  Does the ACL API return a sensible error if you pass a non-existent user?  If so, why not use that?

File: lib/knife-acl/version.rb [hunk 4]
```diff
@@ -1,3 +1,3 @@
 module KnifeACL
-  VERSION = "0.0.12"
+  VERSION = "0.0.13"
```
Comment (stevendanna) [inline lib/knife-acl/version.rb]: We'll need to make this a more significant version change since we are changing the CLI.

**Comment by stevendanna** [comment]
Overall this looks good to me.  There are a few opportunities to reduce some duplication if you wanted, but this is a pretty small code base so I think it is fine.

That said, this is a large pull request.  I would love to get some other eyes on it.  Perhaps @seth or @christophermaier  (past knife-acl contributors).  

If we yank USAG-handling from this code, I think the next step would be to (1) release a new version of the gem, (2) get the code into shape to go into core Chef and then (3) retire this plugin since ACLs are now part of the open source API.  That would likely mean:

- Writing Chef::ACL and Chef::ACL::ACE model classes that handle interacting with the API using the typical methods (load, save, update, etc)

- Tests

File: lib/chef/knife/acl_bulk_add.rb (hunk 1 above)
Comment (jeremiahsnapp) [inline lib/chef/knife/acl_bulk_add.rb]: Thanks for the suggestion. I used your implementation in the latest code.

File: lib/chef/knife/acl_bulk_add.rb (hunk 2 above)
Comment (jeremiahsnapp) [inline lib/chef/knife/acl_bulk_add.rb]: Done

File: lib/knife-acl/version.rb (hunk 4 above)
Comment (jeremiahsnapp) [inline lib/knife-acl/version.rb]: I bumped it to 0.1.0 now.

File: lib/chef/knife/acl_remove.rb (hunk 3 above)
Comment (jeremiahsnapp) [inline lib/chef/knife/acl_remove.rb]: renamed this to `validate_member_exists!`

The API doesn't always return a sensible error.

**Pull request merged**